contracts/protocore/Protocore.sol

```diff
@@ -310,6 +317,216 @@ contract Protocore is MosaicVersion, CoconsensusModule {
+            "Link height inclusion principle has surpassed."
+        );
+
+        address validator = ecrecover(_voteMessageHash, _v, _r, _s);
+
+        require(
+            validator != address(0),
+            "Validator must not be null."
+        );
+
+        bool quorumReached = countVoteForForwardValidatorSets(_voteMessageHash, link, validator);
+        insertForwardValidatorVote(_voteMessageHash, validator);
+
+        if (quorumReached) {
+            justifyLink(link);
+        }
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice calculateQuorum() function calculates quorum of the given
+     *         count.
+     *         The super majority numerator and dominator must be the same
+     *         as in the corresponding core.
+     */
+    function calculateQuorum(uint256 _count)
+        private
+        pure
+        returns (uint256 quorum_)
+    {
+        quorum_ = _count
+            .mul(CORE_SUPER_MAJORITY_NUMERATOR)
+            .div(CORE_SUPER_MAJORITY_DENOMINATOR);
+    }
+
+    function countVoteForForwardValidatorSets(
+        bytes32 _voteMessageHash,
+        Link storage _link,
+        address _validator
+    )
+        private
+        returns (bool quorumReached_)
+    {
+        bool quorumForwardValidatorSet = false;
+        if (canVoteInForwardValidatorSet(
+            _voteMessageHash,
+            openKernelHeight,
+            openKernelHeight,
+            _validator)) {
+            _link.fvsVoteCount[openKernelHeight] = _link.fvsVoteCount[openKernelHeight].add(1);
+            quorumForwardValidatorSet = (_link.fvsVoteCount[openKernelHeight] >= fvsQuorums[openKernelHeight]);
+        }
+
+        bool quorumRearValidatorSet = true;
+        if (openKernelHeight > uint256(0)) {
+            uint256 previousHeight = openKernelHeight.sub(1);
+            if (canVoteInForwardValidatorSet(
+                _voteMessageHash,
+                openKernelHeight,
+                previousHeight,
+                _validator)) {
+                _link.fvsVoteCount[openKernelHeight] = _link.fvsVoteCount[previousHeight].add(1);
+            }
+            quorumRearValidatorSet = (_link.fvsVoteCount[previousHeight] >= fvsQuorums[previousHeight]);
+        }
+        quorumReached_ = (quorumForwardValidatorSet && quorumRearValidatorSet);
+    }
+
+    function canVoteInForwardValidatorSet(
+        bytes32 _voteMessageHash,
+        uint256 _votedHeight,
+        uint256 _fvsHeight,
+        address _validator
+    )
+        private
+        view
+        returns (bool)
+    {
+        bool hasNotVoted = (fvsVotes[_voteMessageHash][_votedHeight][_validator] == address(0));
+        bool inFvs = inForwardValidatorSet(
+            _validator,
+            _fvsHeight
+        );
+        return (hasNotVoted && inFvs);
+    }
+
+    function insertForwardValidatorVote(
+        bytes32 _voteMessageHash,
+        address _validator
+    )
+        private
+    {
+        require(
```

Comment:
Based on our discussion, we can move this into countVoteForForwardValidatorSets() function.

Reply:
I moved it up to `registerVoteInternal()`; seemed sensible to then just assert at the top